### PR TITLE
Feature/user crud

### PR DESCRIPTION
**Pull request by swh-cgm**
User CRUD

PR checklist
https://docs.google.com/spreadsheets/d/16-A1yjOXiQPoTwT-kNkS-5MpvUoznBai-nwQy1y2HFE/edit#gid=444286851

### CODE REVIEW

File: app/Dao/UserDao.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    public function update(array $updateData, int $id): void
+    {
+        $user = User::find($id);
+        $user->name = $updateData['name'];
+        $user->email = $updateData['email'];
+        $user->role = $updateData['role'];
+        if ($updateData['img']) {
+            $user->img = $updateData['img'];
+        }
+        $user->save();
```
Comment (ingyinmaycgm) [inline app/Dao/UserDao.php]: You can use just single code line for this function. 
`User::where('id', $id)->update($updateData);`

File: app/Http/Controllers/UserController.php [hunk 2]
```diff
@@ -4,62 +4,120 @@
...
+        $request->validate([
+            'name' => 'required|max:255',
+            'email' => 'required|email|max:255|unique:users',
+            'password' => 'required|confirmed|min:8|max:255',
+            'password_confirmation' => 'required|min:8|max:255',
+            'img' => 'required|image',
+            'role' => 'required'
+        ]);
+
+        $data = $this->userService->insert($request);
```
Comment (ingyinmaycgm) [inline app/Http/Controllers/UserController.php]: Don't assign to variable if it's not reused. Waste of memory.

File: app/Http/Controllers/UserController.php [hunk 3]
```diff
@@ -4,62 +4,120 @@
...
+    public function update(Request $request, User $user): RedirectResponse
     {
-        //
+        $request->validate([
+            'name' => 'required|max:255',
+            'email' => 'required|email|max:255',
+            'img' => 'image',
+            'role' => 'required'
+        ]);
+        $data = $this->userService->update($request);
```
Comment (ingyinmaycgm) [inline app/Http/Controllers/UserController.php]: Same here. 
`Don't assign to variable if it's not reused. Waste of memory.`

File: app/Http/Controllers/UserController.php [hunk 4]
```diff
@@ -4,62 +4,120 @@
...
      * Remove the specified resource from storage.
+     *
+     * @param integer $id
+     * @return RedirectResponse
      */
-    public function destroy(User $user)
+    public function destroy(int $id): RedirectResponse
     {
-        //
+        $data = $this->userService->delete($id);
```
Comment (ingyinmaycgm) [inline app/Http/Controllers/UserController.php]: Same here. 
`Don't assign to variable if it's not reused. Waste of memory.`

File: app/Services/UserService.php [hunk 5]
```diff
@@ -0,0 +1,106 @@
+<?php
+namespace App\Services;
+
+use App\Dao\UserDao;
```
Comment (ingyinmaycgm) [inline app/Services/UserService.php]: Remove unused class.

File: app/Dao/UserDao.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
+<?php
+namespace App\Dao;
+
+use App\Contracts\Dao\UserDaoInterface;
+use App\Models\User;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Collection;
+use Illuminate\Http\Request;
```
Comment (ingyinmaycgm) [inline app/Dao/UserDao.php]: Remove unused class.

File: app/Services/UserService.php [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        $this->userDao = $userDaoInterface;
+    }
+
+    /**
+     * insert user
+     *
+     * @param Request $request
+     * @return null
+     */
+    public function insert(Request $request): null
```
Comment (ingyinmaycgm) [inline app/Services/UserService.php]: May be the return type is `void` if you don't return anything.

File: app/Services/UserService.php [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    {
+        $encrypted = Hash::make($request->password);
+
+        // $path = $request->file('img')->store('pfp');
+
+        $filename = $request->img->getClientOriginalName();
```
Comment (ingyinmaycgm) [inline app/Services/UserService.php:30]: ~~Line numbers 33 ~ 43 should be in service layer.~~
And may be you don't need to return for this function. UserDao->insert() doesn't return anything.

File: app/Services/UserService.php [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+     * insert user
+     *
+     * @param Request $request
+     * @return null
+     */
+    public function insert(Request $request): null
+    {
+        $encrypted = Hash::make($request->password);
+
+        // $path = $request->file('img')->store('pfp');
```
Comment (ingyinmaycgm) [inline app/Services/UserService.php]: Remove this if you don't use.

File: app/Services/UserService.php [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        return $data;
+    }
+
+    /**
+     * Delete user
+     *
+     * @param int $id
+     * @return null
+     */
+    public function delete(int $id): null
```
Comment (ingyinmaycgm) [inline app/Services/UserService.php]: The return type should be void and you don't need to return anything for this. There is no return value from UserDao->delete().

File: app/Services/UserService.php [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        return $this->userDao->delete($id);
+    }
+
+    /**
+     * update user
+     *
+     * @param Request $request
+     * @return null
+     */
+    public function update(Request $request): null
```
Comment (ingyinmaycgm) [inline app/Services/UserService.php]: Same here. 
https://github.com/swh-cgm/laravel-OJT/pull/2#discussion_r1299562173

File: composer.json [hunk 12]
```diff
@@ -26,7 +26,10 @@
             "App\\": "app/",
             "Database\\Factories\\": "database/factories/",
             "Database\\Seeders\\": "database/seeders/"
-        }
+        },
+        "files": [
```
Comment (ingyinmaycgm) [inline composer.json]: May be you don't need this code line.

File: resources/views/users/index.blade.php [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+  <td>{{ $user->email }}</td>
+  <td>{{ $user->name }}</td>
+  <td><a href="{{ route('users.detail', $user->id) }}" rel="User Detail">User Details</td>
+  <td><a href="{{ route('users.edit', $user->id) }}" rel="Edit User">Edit User</td>
+  <td><a class="text-danger" href="{{ route('users.delete', $user->id) }}" rel="Delete User">Delete User</td>
+</tr>
+@endforeach
+</table>
+
+<a>
```
Comment (ingyinmaycgm) [inline resources/views/users/index.blade.php]: For what?

File: resources/views/users/index.blade.php [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+<table class="user-table">
+  <tr>
+    <th>User ID</th>
+    <th>User Email</th>
+    <th>User Name</th>
+    <th>User Detail</th>
+    <th>Edit User</th>
+    <th>Delete User</th>
+  </tr>
+@foreach ($users as $user)
```
Comment (ingyinmaycgm) [inline resources/views/users/index.blade.php:23]: Fix indentation. 

> 
```blade

``` <table>
  <tr>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
  </tr>
</table>

File: routes/web.php [hunk 15]
```diff
@@ -16,3 +19,15 @@
 Route::get('/', function () {
     return view('welcome');
 });
+
+// Route::get('/create', function () {
```
Comment (ingyinmaycgm) [inline routes/web.php]: Remove comment.

File: routes/web.php [hunk 16]
```diff
@@ -1,6 +1,9 @@
 <?php
 
 use Illuminate\Support\Facades\Route;
+use App\Http\Controllers\UserController;
+use App\Http\Controllers\FileDownloadController;
```
Comment (ingyinmaycgm) [inline routes/web.php]: Remove unused class.

File: app/Services/UserService.php (hunk 7 above)
Comment (swh-cgm) [reply app/Services/UserService.php]: when return type is changed to null, it throws the following error
"A void function must not return a value"

at 44
return $this->userDao->insert($insertData);

**Pull request merged**